shuffle-metrics-reporter/shuffle-metrics-reporter-prometheus/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/AbstractPrometheusReporter.java

```diff
@@ -0,0 +1,346 @@
+import com.alibaba.metrics.MetricRegistry;
+import com.alibaba.metrics.Snapshot;
+import com.alibaba.metrics.Timer;
+import com.alibaba.metrics.common.config.MetricsCollectPeriodConfig;
+import com.alibaba.metrics.reporter.MetricManagerReporter;
+import io.prometheus.client.Collector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Pattern;
+
+/** base prometheus reporter for prometheus metrics. */
+public class AbstractPrometheusReporter extends MetricManagerReporter {
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+    private static final Pattern UNALLOWED_CHAR_PATTERN = Pattern.compile("[^a-zA-Z0-9:_]");
+    private IMetricManager metricManager;
+
+    private final Map<String, AbstractMap.SimpleImmutableEntry<Collector, Integer>>
+            collectorsWithCountByMetricName = new ConcurrentHashMap<>();
+
+    public AbstractPrometheusReporter(
+            IMetricManager metricManager,
+            String name,
+            MetricFilter filter,
+            MetricsCollectPeriodConfig metricsReportPeriodConfig,
+            TimeUnit rateUnit,
+            TimeUnit durationUnit) {
+        super(metricManager, name, filter, metricsReportPeriodConfig, rateUnit, durationUnit);
+        this.metricManager = metricManager;
+    }
+
+    @Override
+    public void report(
+            Map<MetricName, Gauge> gauges,
+            Map<MetricName, Counter> counters,
+            Map<MetricName, Histogram> histograms,
+            Map<MetricName, Meter> meters,
+            Map<MetricName, Timer> timers,
+            Map<MetricName, Compass> compasses,
+            Map<MetricName, FastCompass> fastCompasses,
+            Map<MetricName, ClusterHistogram> clusterHistogrames) {
+
+        Map<String, Set<MetricName>> metricNamesByGroup = metricManager.listMetricNamesByGroup();
+
+        Arrays.asList(
+                        gauges,
+                        counters,
+                        histograms,
+                        meters,
+                        timers,
+                        compasses,
+                        fastCompasses,
+                        clusterHistogrames)
+                .forEach(
+                        metricNameMap -> {
+                            for (Map.Entry<MetricName, ? extends Metric> g :
+                                    metricNameMap.entrySet()) {
+
+                                String group = getGroupByMetricName(metricNamesByGroup, g.getKey());
+                                String metricName =
+                                        filterCharacters(
```

Comment:
It seems we will always filter the invalid character, if so, should we remove the FILTER_LABEL_VALUE_CHARACTER config option?